pkg/discovery/repository/kube_cluster.go

```diff
@@ -33,76 +34,144 @@ type KubeCluster struct {
-	MountName  string
+func CreateClusterSummary(kubeCluster *KubeCluster) *ClusterSummary {
+	clusterSummary := &ClusterSummary{
+		KubeCluster:              kubeCluster,
+		NodeMap:                  make(map[string]*KubeNode),
+		NodeNameUIDMap:           make(map[string]string),
+		NamespaceUIDMap:          make(map[string]string),
+		PodClusterIDToServiceMap: make(map[string]*v1.Service),
+		NodeToRunningPods:        make(map[string][]*v1.Pod),
+		NodeToPendingPods:        make(map[string][]*v1.Pod),
+		ClusterResources:         make(map[metrics.ResourceType]*KubeDiscoveredResource),
+	}
+	clusterSummary.computeNodeMap()
+	clusterSummary.computePodMap()
+	clusterSummary.computeQuotaMap()
+	clusterSummary.computePodToServiceMap()
+	clusterSummary.computeClusterResources()
+	return clusterSummary
 }
 
-// Volume metrics reported for a given pod
-type PodVolumeMetrics struct {
-	Volume   *v1.PersistentVolume
-	Capacity float64
-	Used     float64
-	PodVolume
+func (summary *ClusterSummary) GetKubeNamespace(namespace string) *KubeNamespace {
+	kubeNamespace, exists := summary.NamespaceMap[namespace]
+	if !exists {
+		glog.Errorf("Cannot find KubeNamespace entity for namespace %s", namespace)
+		return nil
+	}
+	return kubeNamespace
 }
 
-func NewKubeCluster(clusterName string, nodes []*v1.Node) *KubeCluster {
-	kubeCluster := &KubeCluster{
-		Name:             clusterName,
-		Nodes:            make(map[string]*KubeNode),
-		Namespaces:       make(map[string]*KubeNamespace),
-		ClusterResources: make(map[metrics.ResourceType]*KubeDiscoveredResource),
+func (summary *ClusterSummary) GetReadyPods() (readyPods []*v1.Pod) {
+	for _, pod := range summary.Pods {
+		if util.PodIsReady(pod) {
+			readyPods = append(readyPods, pod)
+		}
 	}
-	kubeCluster.addNodes(nodes)
-	if glog.V(3) {
-		kubeCluster.logClusterNodes()
+	return
+}
+
+func (summary *ClusterSummary) GetRunningPodsOnNode(node *v1.Node) []*v1.Pod {
+	runningPods, exists := summary.NodeToRunningPods[node.Name]
+	if !exists {
+		glog.V(3).Infof("No running pods found on node %v.", node.Name)
+		return nil
 	}
-	kubeCluster.computeClusterResources()
-	return kubeCluster
+	return runningPods
+
 }
 
-func (kc *KubeCluster) addNodes(nodes []*v1.Node) *KubeCluster {
-	for _, node := range nodes {
-		// Create kubeNode with Allocatable resource as the
-		// resource capacity
-		kc.Nodes[node.Name] = NewKubeNode(node, kc.Name)
+func (summary *ClusterSummary) GetPendingPodsOnNode(node *v1.Node) []*v1.Pod {
+	pendingPods, exists := summary.NodeToPendingPods[node.Name]
+	if !exists {
+		glog.V(3).Infof("No pending pods found on node %v.", node.Name)
+		return nil
 	}
-	return kc
+	return pendingPods
+
+}
+
+func (summary *ClusterSummary) computePodMap() {
+	for _, pod := range summary.Pods {
+		nodeName := pod.Spec.NodeName
+		if nodeName == "" {
+			glog.V(3).Infof("Skip pod %v/%v without assigned node.",
+				pod.Namespace, pod.Name)
+			continue
+		}
+		switch pod.Status.Phase {
+		case v1.PodPending:
```

Comment:
Repeating @chlam4's comment here. `nodeName` would be `""` for pending pods.

Reply:
As explained above, a pod bound to a node can still have a `Pending` phase.